apps/web/src/hooks/usePermit2Allowance.ts

```diff
@@ -0,0 +1,152 @@
+import { useInterval } from '@pancakeswap/hooks'
+import { Permit2Signature } from '@pancakeswap/universal-router-sdk'
+import { CurrencyAmount, Token } from '@pancakeswap/sdk'
+import { usePermitAllowance, useUpdatePermitAllowance } from 'hooks/usePermitAllowance'
+import { Dispatch, SetStateAction, useCallback, useMemo, useState } from 'react'
+import { useHasPendingApproval, useHasPendingRevocation } from 'state/transactions/hooks'
+import { useAccount, Address } from 'wagmi'
+import { ApprovalState, useApproveCallback } from './useApproveCallback'
+import useTokenAllowance from './useTokenAllowance'
+
+export enum AllowanceState {
+  LOADING,
+  REQUIRED,
+  ALLOWED,
+}
+
+export interface BaseAllowance {
+  state: AllowanceState
+  setSignature: Dispatch<SetStateAction<Permit2Signature | undefined>>
+  [k: string]: unknown
+}
+
+export interface AllowanceLoading extends BaseAllowance {
+  state: AllowanceState.LOADING
+}
+export interface AllowanceAllowed extends BaseAllowance {
+  state: AllowanceState.ALLOWED
+  permitSignature?: Permit2Signature
+}
+
+export interface AllowanceRequired extends BaseAllowance {
+  isRevocationPending: boolean
+  state: AllowanceState.REQUIRED
+  token: Token
+  isApprovalLoading: boolean
+  isApprovalPending: boolean
+  approve: any
+  permit: () => Promise<void>
+  revoke: any
+  needsSetupApproval: boolean
+  needsPermitSignature: boolean
+  allowedAmount: CurrencyAmount<Token>
+}
+
+export type Allowance = AllowanceLoading | AllowanceAllowed | AllowanceRequired
+
+const AVERAGE_L1_BLOCK_TIME = 12000
+
+export default function usePermit2Allowance(
+  approvalAddress: string | undefined,
+  amount?: CurrencyAmount<Token>,
+  spender?: Address,
+): Allowance {
+  const { address: account } = useAccount()
+  const token = amount?.currency
+
+  const { allowance: tokenAllowance } = useTokenAllowance(token, account, approvalAddress)
```

Comment:
shouldn't we check for the allowance of the permit2 contract?

Reply:
why? we need check the allowance to permit2 address, and the permit amount

Reply:
what's the `approvalAddress` here? The approval address should be the address of permit2, not some arbitrary address passed in from the hook

Reply:
we first need to check if we have enough allowance set for the permit2 contract

Reply:
yes, the previous support wallchain, so the address may wallchain permit address
but now we disabled wallchain, so the address will always permit2 address

Reply:
hmm alright. Let's remove the approvalAddress here. Ideally should ask them to integrate with our permit2 instead of checking approval here

Reply:
once they're integrated with our permit2 then should be following the same flow